src/write/xcoff.rs

```diff
@@ -0,0 +1,489 @@
+                        }
+                    }
+                    SymbolKind::Tls => {
+                        if symbol.is_local() {
+                            xcoff::C_STTLS
+                        } else {
+                            xcoff::C_GTLS
+                        }
+                    }
+                    SymbolKind::Section => {
+                        if symbol.weak {
+                            xcoff::C_WEAKEXT
+                        } else if symbol.is_undefined() {
+                            xcoff::C_HIDEXT
+                        } else {
+                            xcoff::C_EXT
+                        }
+                    }
+                    SymbolKind::Unknown => {
+                        return Err(Error(format!(
+                            "unimplemented symbol `{}` kind {:?}",
+                            symbol.name().unwrap_or(""),
+                            symbol.kind
+                        )));
+                    }
+                }
+            };
+            let sym_type = if (symbol.scope == SymbolScope::Linkage)
+                && (storage_class == xcoff::C_EXT
+                    || storage_class == xcoff::C_WEAKEXT
+                    || storage_class == xcoff::C_HIDEXT)
+            {
+                xcoff::SYM_V_HIDDEN
+            } else {
+                0
+            };
+            let aux_num = symbol_offsets[index].aux_count;
+            if is_64 {
+                let xcoff_sym = xcoff::Symbol64 {
+                    n_value: U64::new(BE, symbol.value),
+                    n_offset: U32::new(
+                        BE,
+                        strtab.get_offset(symbol_offsets[index].str_id.unwrap()) as u32,
+                    ),
+                    n_scnum: I16::new(BE, section_number),
+                    n_type: U16::new(BE, sym_type),
+                    n_sclass: storage_class,
+                    n_numaux: aux_num,
+                };
+                buffer.write(&xcoff_sym);
+            } else {
+                let mut sym_name = [0; 8];
+                let name = &symbol.name[..];
+                if name.len() <= 8 {
+                    sym_name[..name.len()].copy_from_slice(name);
+                } else {
+                    let str_offset = strtab.get_offset(symbol_offsets[index].str_id.unwrap());
+                    sym_name[4..8].copy_from_slice(&u32::to_be_bytes(str_offset as u32));
+                }
+                let xcoff_sym = xcoff::Symbol32 {
+                    n_name: sym_name,
+                    n_value: U32::new(BE, symbol.value as u32),
+                    n_scnum: I16::new(BE, section_number),
+                    n_type: U16::new(BE, sym_type),
+                    n_sclass: storage_class,
+                    n_numaux: aux_num,
+                };
+                buffer.write(&xcoff_sym);
+            }
+            // Generate a csect auxiliary entry.
+            if symbol.kind == SymbolKind::Section {
```

Comment:
I'm concerned about making symbols that have csects a `SymbolKind::Section` symbol. For current implementation, the name of `SymbolKind::Section` symbol is not recorded and finally the name is an empty string in the symbol table. We can either
1. Change `add_symbol` function to allow the name of `SymbolKind::Section` symbol being recorded for xcoff and set symbol's flags to `SymbolFlags::Xcoff {...}` when reading the object file.
2. Map object's predefined symbol kinds to xcoff's storage classes and create csects by storage classses.

Reply:
> I'm concerned about making symbols that have csects a SymbolKind::Section symbol.

I agree.  As per my previous request, this can't be correct and I would like it changed before merging this PR.

Is it possible to do both of those? That is, use `SymbolFlags::Xcoff` if it is present, and otherwise use a predefined mapping.

Reply:
> Is it possible to do both of those? That is, use SymbolFlags::Xcoff if it is present, and otherwise use a predefined mapping.

That sounds reasonable to me. @EsmeYi What do you think?

Reply:
Sounds reasonable. Since the variable `storage_class` has been obtained by this rule, it can be realized by checking this variable here.
( I should have changed the logic here when adding `SymbolFlags::Xcoff`, thanks for pointing it out.)

Reply:
> the name of SymbolKind::Section symbol is not recorded and finally the name is an empty string in the symbol table

@bzEq Does this look ok to you now? I'm not actually sure which name you are referring to, so please check this.

Reply:
Updated code for handling SymbolKind looks good to me now.

> I'm not actually sure which name you are referring to, so please check this.

```rust
    /// Add a new symbol and return its `SymbolId`.                                                                                               
    pub fn add_symbol(&mut self, mut symbol: Symbol) -> SymbolId {                                                                                
        // Defined symbols must have a scope.                                                                                                     
        debug_assert!(symbol.is_undefined() || symbol.scope != SymbolScope::Unknown);                                                             
        if symbol.kind == SymbolKind::Section {                                                                                                   
            // There can only be one section symbol, but update its flags, since                                                                  
            // the automatically generated section symbol will have none.                                                                         
            let symbol_id = self.section_symbol(symbol.section.id().unwrap());                                                                    
            if symbol.flags != SymbolFlags::None {                                                                                                
                self.symbol_mut(symbol_id).flags = symbol.flags;                                                                                  
            }                                                                                                                                     
            return symbol_id;                                                                                                                     
        }
...
    /// Get the symbol for a section.                                                                                                             
    pub fn section_symbol(&mut self, section_id: SectionId) -> SymbolId {                                                                         
        let section = &mut self.sections[section_id.0];                                                                                           
        if let Some(symbol) = section.symbol {                                                                                                    
            return symbol;                                                                                                                        
        }                                                                                                                                         
        let name = if self.format == BinaryFormat::Coff {                                                                                         
            section.name.clone()                                                                                                                  
        } else {                                                                                                                                  
            Vec::new()                                                                                                                            
        };                                                                                                                                        
        let symbol_id = SymbolId(self.symbols.len());                                                                                             
        self.symbols.push(Symbol {                                                                                                                
            name,
```
When adding a section symbol, the name is only recorded for coff currently.